# Cryptic transmission of SARS-CoV-2 in Washington State

Trevor Bedford<sup>1,2,3</sup>, Alexander L. Greninger<sup>1,3</sup>, Pavitra Roychoudhury<sup>1,3</sup>, Lea M. Starita<sup>2,3</sup>, Michael Famulare<sup>4</sup>, Meei-Li Huang<sup>3</sup>, Arun Nalla<sup>3</sup>, Gregory Pepper<sup>3</sup>, Adam Reinhardt<sup>3</sup>, Hong Xie<sup>3</sup>, Lasata Shrestha<sup>3</sup>, Truong N Nguyen<sup>3</sup>, Amanda Adler<sup>5</sup>, Elisabeth Brandstetter<sup>3</sup>, Shari Cho<sup>3</sup>, Danielle Giroux<sup>3</sup>, Peter D. Han<sup>3</sup>, Kairsten Fay<sup>1</sup>, Chris D. Frazar<sup>3</sup>, Misja Ilcisin<sup>1</sup>, Kirsten Lacombe<sup>5</sup>, Jover Lee<sup>1</sup>, Anahita Kiavand<sup>3</sup>, Matthew Richardson<sup>3</sup>, Thomas R. Sibley<sup>1</sup>, Melissa Truong<sup>3</sup>, Caitlin R. Wolf<sup>3</sup>, Deborah A. Nickerson<sup>2,3</sup>, Mark J. Rieder<sup>2,3</sup>, Janet A. Englund<sup>3,5</sup>, the Seattle Flu Study Investigators, James Hadfield<sup>1</sup>, Emma B. Hodcroft<sup>6</sup>, John Huddleston<sup>1,3</sup>, Louise H. Moncla<sup>1</sup>, Nicola F. MÃ¼ller<sup>1</sup>, Richard A. Neher<sup>6</sup>, Xianding Deng<sup>7</sup>, Wei Gu<sup>7</sup>, Scot Federman<sup>7</sup>, Charles Chiu<sup>7</sup>, Jeff Duchin<sup>3,8</sup>, Romesh Gautom<sup>9</sup>, Geoff Melly<sup>9</sup>, Brian Hiatt<sup>9</sup>, Philip Dykema<sup>9</sup>, Scott Lindquist<sup>9</sup>, Krista Queen<sup>10</sup>, Ying Tao<sup>10</sup>, Anna Uehara<sup>10</sup>, Suxiang Tong<sup>10</sup>, Duncan MacCannell<sup>10</sup>, Gregory L. Armstrong<sup>10</sup>, Geoffrey S. Baird<sup>3</sup>, Helen Y. Chu<sup>2,3</sup>, Jay Shendure<sup>2,3,11</sup>, Keith R. Jerome<sup>1,3</sup>

# Abstract

Following its emergence in Wuhan, China, in late November or early December 2019, the SARS-CoV-2 virus has rapidly spread throughout the world. On March 11, 2020, the World Health Organization declared Coronavirus Disease 2019 (COVID-19) a pandemic. Genome sequencing of SARS-CoV-2 strains allows for the reconstruction of transmission history connecting these infections. Here, we analyze 346 SARS-CoV-2 genomes from samples collected between 20 February and 15 March 2020 from infected patients in Washington State, USA. We found that the large majority of SARS-CoV-2 infections sampled during this time frame appeared to have derived from a single introduction event into the state in late January or early February 2020 and subsequent local spread, strongly suggesting cryptic spread of COVID-19 during the months of January and February 2020, before active community surveillance was implemented. We estimate a common ancestor of this outbreak clade as occurring between 18 January and 9 February 2020. From genomic data, we estimate an exponential doubling between 2.4 and 5.1 days. These results highlight the need for large-scale community surveillance for SARS-CoV-2 introductions and spread and the power of pathogen genomics to inform epidemiological understanding.

# Organization

* `augur-build`: Contains consensus SARS-CoV-2 genomes in `data/` and a reproducible Nextstrain phylogenetic pipeline. Results of this pipeline are viewable at [nextstrain.org/community/blab/ncov-cryptic-transmission/wa1-clade](https://nextstrain.org/community/blab/ncov-cryptic-transmission/wa1-clade) (Figure 1a).
* `beast`: Contains XML with genome data, BEAST output and a Mathematica analysis notebook (Figure 1b)
* `forward-sim`: Contains MatLab code to conduct forward simulation (Figure 3)
* Figure 2 and Supp Figure 1 are generated from the Nextstrain `ncov` build run on 25 March 2020 available at [nextstrain.org/ncov/2020-03-25](https://nextstrain.org/ncov/2020-03-25). The code to reproduce these figures is available at commit [3ac03a9a](https://github.com/nextstrain/ncov/tree/3ac03a9a80dc4dd28aa8e44dd1312e6531f3a6ea).
